File: Graph/Union-Find/programmers_섬 연결하기.kt
//https://programmers.co.kr/learn/courses/30/lessons/42861
class Solution {
    fun getParent(parent : IntArray, x : Int): Int{
        return if(parent[x]==x) x else getParent(parent,parent[x]).also{ parent[x] = it}
    }
   fun unionParent(parent : IntArray, a : Int, b : Int){
       val x = getParent(parent, a)
       val y = getParent(parent, b)
       if(x<y){
           parent[y]=x
       }
       else{
           parent[x]=y
       }
   }
   fun findParent(parent : IntArray, a : Int, b : Int) : Boolean{
       val x = getParent(parent, a)
       val y = getParent(parent, b)
       return if(x==y) true else false
    }
    
    fun solution(n: Int, costs: Array<IntArray>): Int {
        var answer = 0
        val parent = IntArray(n){it}
        //비용 별로 오름차순
        costs.sortWith(Comparator(){a,b ->a[2]-b[2]})

        for(i in costs.indices){
            if(!findParent(parent,costs[i][0],costs[i][1])){
                //부모가 같지 않다면 (싸이클이 없다면)
                unionParent(parent,costs[i][0],costs[i][1])
                answer+=costs[i][2]
            }
        }
        return answer
    }
}

File: Graph/Union-Find/programmers_섬 연결하기.cpp
//https://programmers.co.kr/learn/courses/30/lessons/42861
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int getParent(vector<int> &parent, int x){
    if(parent[x]==x) return x;
    return parent[x] = getParent(parent, parent[x]);
}

void unionParent(vector<int> &parent, int a, int b){
    a = getParent(parent, a);
    b = getParent(parent, b);
    if( a<b) parent[b] = a;
    else parent[a] = b;
}

bool findParent(vector<int> &parent, int a, int b){
    a = getParent(parent, a);
    b = getParent(parent, b);
    if(a==b) return 1;
    else return 0;
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
vector<pair<int,pair<int,int>>> vt;
    vector<int> parent;
    for(int i=0; i<n;i++){
        parent.push_back(i);
    }
    for(int i=0; i<costs.size();i++){
        vt.push_back({costs[i][2],{costs[i][0],costs[i][1]}});
    }
    sort(vt.begin(),vt.end());
    for(int i=0; i< vt.size();i++){

      if(!findParent(parent, vt[i].second.first, vt[i].second.second)){
          answer+= vt[i].first;
          unionParent(parent, vt[i].second.first, vt[i].second.second);
      }
    }

    return answer;
}
